### PR TITLE
Migrated from miniconda to official python bullseye docker image

### LINKED ISSUE #3666: Reduce the number of Docker Images to just one image
Currently we build and push 5 variants of Docker Images regularly i.e. `latest`, `jax`, `odes`, `idaklu` & `all`.
The idea behind this issue is to have a single standard image variant to be updated regularly to save resources and make things simple for developers.
As discussed in #3661

### LINKED ISSUE #3692: Migrate from `conda` to `venv` inside Docker Images
This would potentially cut down an external system dependency `conda` & would reduce the inconsistency of environments across different OS for IDAKLU solver images.

### PR DESCRIPTION
**Pull request by santacodes**
# Description

This draft is in continuation to #3874 which was closed due to messy branches and the image failing to build IDAKLU solver on aarch64 linux. Initially the idea was to migrate to a single docker image of manylinux2014 which was compliant to [PEP599](https://peps.python.org/pep-0599/), but due to some complications, the official python image along with venv were decided upon implementation, check #3874 discussions for more information.

Fixes #3692
Fixes #3666 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Thanks @santacodes, happy to keep this as a draft for now for visibility. xref #3879 which needs to be resolved before we are able to come back to this.

**Comment by cringeyburger** [comment]
Hey, @santacodes! I wanted to ask if you could finish the pull request. 
I need to make some changes to the Dockerfile to adjust the build commands, but since we're planning to move away from `conda-forge`, it would be great if you could finalize this PR and get it merged. 
This will allow the migration PR to move forward smoothly.

@arjxn-py has also agreed to help review the PR.
Thanks a lot!

**Comment by santacodes** [comment]
> Hey, @santacodes! I wanted to ask if you could finish the pull request. I need to make some changes to the Dockerfile to adjust the build commands, but since we're planning to move away from `miniconda`, it would be great if you could finalize this PR and get it merged. This will allow the migration PR to move forward smoothly.
> 
> @arjxn-py has also agreed to help review the PR. Thanks a lot!

Sure I'll get on it!

**Comment by santacodes** [comment]
@arjxn-py do you mind if I change the standard `pip` installations of packages to `uv`? Could it possibly reduce the build times?

**Comment by arjxn-py** [comment]
I don't mind at all @santacodes, feel free to switch to `uv`. We also have an open issue for the same #3825, this would be a good step in that direction.

**Comment by arjxn-py** [comment]
You'd also want to set the docker workflow to temporarily run on this PR

**Comment by santacodes** [comment]
Also while running unit tests inside the docker container it promoted this - 
```bash 
SKIPPED [1] tests/unit/test_solvers/test_idaklu_jax.py:91: Both IDAKLU and JAX are available
SKIPPED [1] tests/unit/test_util.py:91: The JAX solver is not installed
```
I don't know if it has been addressed already but I see the same thing on the CI as well, I think this part of code in `tests/unit/test_util.py` is causing it, instead of "The JAX solver is not installed" it should be "The JAX solver is installed". Correct me if I am wrong 😅 .
```python
@pytest.mark.skipif(pybamm.have_jax(), reason="The JAX solver is not installed")
    def test_is_jax_compatible(self):
        assert True

```

**Comment by arjxn-py** [comment]
Yes currently it skips if `pybamm.have_jax()`, which should not happen.
I should have been a mistake while #4214, see [here](https://github.com/pybamm-team/PyBaMM/pull/4214/files#diff-213a36200d41e6897aaba82ed856a037dcf622587acf920011b95c64d3428338R76-R78) 
Feel free to modify this, there should not be a condition to skip and it should check for `pybamm.is_jax_compatible()`. Hope i'm not missing anything.

**Comment by arjxn-py** [comment]
> I noticed the workflow failed because of a login error, should

Feel free to skip login & switch to cli build command temporarily: 
```
docker buildx build \
  --file scripts/Dockerfile \
  --tag pybamm/pybamm:latest \
  --push \
  --platform linux/amd64,linux/arm64 \
  .
```
if that works.

> Should I just fix it in this PR or make a new one? Also, the assert inside the function should be False if I am not wrong?

assert should be true after `pybamm.is_jax_compatible()` as if `is_jax_compatible()` returns false, it should fail. Separate PR should be good as it's unrelated to this one.

File: scripts/Dockerfile [hunk 1]
```diff
@@ -21,17 +21,20 @@ ENV CMAKE_CXX_COMPILER=/usr/bin/g++
 ENV CMAKE_MAKE_PROGRAM=/usr/bin/make
 ENV LD_LIBRARY_PATH=/home/pybamm/.local/lib
 
-RUN conda create -n pybamm python=3.11
-RUN conda init --all
-SHELL ["conda", "run", "-n", "pybamm", "/bin/bash", "-c"]
-RUN conda install -y pip
+# Migrating to venv and create a virtual environment for python 3.12
```
Comment (kratman) [inline scripts/Dockerfile:24]:
```suggestion
# Create a virtual environment
```

File: scripts/Dockerfile [hunk 2]
```diff
@@ -21,17 +21,20 @@ ENV CMAKE_CXX_COMPILER=/usr/bin/g++
...
+RUN python3.12 -m venv $VIRTUAL_ENV
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
 
-RUN pip install --upgrade --user pip setuptools wheel wget
-RUN pip install cmake
+RUN pip install uv
+RUN uv pip install --upgrade setuptools wheel wget && \
+uv pip install cmake
+RUN #!/bin/bash && \
+source /home/pybamm/venv/bin/activate;
```
Comment (kratman) [inline scripts/Dockerfile:33]: Can we just do it this way to make it look a bit simpler?
```suggestion
RUN pip install uv
RUN uv pip install --upgrade setuptools wheel wget cmake
RUN #!/bin/bash && source /home/pybamm/venv/bin/activate;
```

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -2,6 +2,7 @@ name: Build and push Docker images to Docker Hub
 
 on:
   workflow_dispatch:
+  pull_request:
```
Comment (kratman) [inline .github/workflows/docker.yml:5]: Don't forget to revert this, we don't want every PR pushed to docker right?

Comment (arjxn-py) [reply .github/workflows/docker.yml:5]: Yes, it's just temporarily changed to test on this PR

Comment (santacodes) [reply .github/workflows/docker.yml:5]: Yes just testing if the image builds in this PR

Comment (kratman) [reply .github/workflows/docker.yml:5]: I think the build completed (and it works locally for me), just the login failed before you made the other changes to the file.

**Comment by kratman** [comment]
@santacodes Just an FYI, force pushing removes the history of what was reviewed. Try not to force push after review has started

**Comment by santacodes** [comment]
> @santacodes Just an FYI, force pushing removes the history of what was reviewed. Try not to force push after review has started

Sorry my commits got messed up between 2 branches, so I had to rebase them 😅 will keep that in mind!

File: .github/workflows/docker.yml (hunk 3 above)
Comment (santacodes) [reply .github/workflows/docker.yml:5]: The image builds locally for me as well, the problem seems to be in the CI command.

Comment (kratman) [reply .github/workflows/docker.yml:5]: Due to the force push I can't easily get back to it, but to me it looked like the CI did finish before, it just failed to push to dockerhub

Comment (santacodes) [reply .github/workflows/docker.yml:5]: Have a look [here](https://github.com/pybamm-team/PyBaMM/actions/runs/10285168300/job/28462799902?pr=3901), these are the logs from the most recent push.

Comment (arjxn-py) [reply .github/workflows/docker.yml:5]: It fails on build, can u try with simple `docker build -t pybamm:latest -f scripts/Dockerfile .` once

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -23,20 +24,13 @@ jobs:
...
-          tags: pybamm/pybamm:latest
-          push: true
-          platforms: linux/amd64, linux/arm64
+      - name: Build docker images
+        run: docker buildx build \
+            --file scripts/Dockerfile \
+            --tag pybamm/pybamm:latest \
+            --push \
+            --platform linux/amd64,linux/arm64 \
+            .
```
Comment (arjxn-py) [inline .github/workflows/docker.yml:33]:
```suggestion
        run: |
          docker build -t pybamm:latest -f scripts/Dockerfile .
```

File: .github/workflows/docker.yml (hunk 3 above)
Comment (santacodes) [reply .github/workflows/docker.yml:5]: Seems to be working now 😃

Comment (arjxn-py) [reply .github/workflows/docker.yml:5]: Nah you're good, it's a little confusing as `docker buildx` requires login for build as well. I missed it before.

Comment (kratman) [reply .github/workflows/docker.yml:5]: Ok revert the workflow and we can get it approved and squashed

**Review by kratman — APPROVED** [review]

**Review by arjxn-py — APPROVED** [review]
Thanks @santacodes

File: scripts/Dockerfile [hunk 5]
```diff
@@ -21,17 +21,18 @@ ENV CMAKE_CXX_COMPILER=/usr/bin/g++
...
 ENV LD_LIBRARY_PATH=/home/pybamm/.local/lib
 
-RUN conda create -n pybamm python=3.11
-RUN conda init --all
-SHELL ["conda", "run", "-n", "pybamm", "/bin/bash", "-c"]
-RUN conda install -y pip
+# Create a virtual environment
+ENV VIRTUAL_ENV=/home/pybamm/venv
+RUN python3.12 -m venv $VIRTUAL_ENV
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
```
Comment (agriyakhetarpal) [inline scripts/Dockerfile:27]: If we are going to use `uv`, could we let `uv` create the virtual environment using `uv venv`?

Comment (santacodes) [reply scripts/Dockerfile]: For context, I am using [these](https://docs.astral.sh/uv/guides/integration/docker/#running-in-docker) procedures listed on `astral-uv` documentation to use `uv` inside the docker container in the latest commit.

**Review by agriyakhetarpal — APPROVED** [review]

**Pull request merged**